### PR TITLE
Add `"autoreject_local"` as new option for `preprocessing/ptp_reject`

### LINKED ISSUE #410: Add support for AutoReject local
With Tomas D'Amelio we implemented the AutoReject local option in ptp_reject to clean the DEAP data. Results are pretty nice! Some work is necessary to make that code fit into master after all we merged lately but we should do it.

### PR DESCRIPTION
**Pull request by hoechenberger**
### Before merging …

- [x] Changelog has been updated (`docs/source/changes.md`)

closes #410

cc @dengemann

### CODE REVIEW

File: mne_bids_pipeline/steps/preprocessing/_08_ptp_reject.py [hunk 1]
```diff
@@ -79,46 +79,65 @@ def drop_ptp(
...
+            n_interpolate=np.array(cfg.autoreject_n_interpolate),
+            random_state=cfg.random_state,
+            n_jobs=exec_params.n_jobs,
+        )
+        ar.fit(epochs)
+        n_epochs_before_reject = len(epochs)
+        epochs, reject_log = ar.transform(epochs, return_log=True)
+        n_epochs_after_reject = len(epochs)
+    else:
+        reject = _get_reject(
```
Comment (hoechenberger) [inline mne_bids_pipeline/steps/preprocessing/_08_ptp_reject.py:100]: The diff doesn't look nice here, but I basically just added this new `if` block and moved the previous code into the `else` section

File: mne_bids_pipeline/_config.py [hunk 2]
```diff
@@ -1405,13 +1401,28 @@
 reject_tmax: Optional[float] = None
 """
 End of the time window used to reject epochs. If `None`, the window will end
-with the last time point.
+with the last time point.Has no effect if
```
Comment (hoechenberger) [inline mne_bids_pipeline/_config.py:1404]:
```suggestion
with the last time point. Has no effect if
```

File: mne_bids_pipeline/steps/preprocessing/_08_ptp_reject.py [hunk 3]
```diff
@@ -79,46 +79,65 @@ def drop_ptp(
...
+        logger.info(**gen_log_kwargs(message=msg))
+
+        ar = autoreject.AutoReject(
+            n_interpolate=np.array(cfg.autoreject_n_interpolate),
+            random_state=cfg.random_state,
+            n_jobs=exec_params.n_jobs,
+        )
+        ar.fit(epochs)
+        n_epochs_before_reject = len(epochs)
+        epochs, reject_log = ar.transform(epochs, return_log=True)
```
Comment (hoechenberger) [inline mne_bids_pipeline/steps/preprocessing/_08_ptp_reject.py:97]:
```suggestion
        n_epochs_before_reject = len(epochs)
        epochs, reject_log = ar.fit_transform(epochs, return_log=True)
```

**Comment by hoechenberger** [comment]
@larsoner I'm not sure how to modify `_config_import.py` in order to fix this test:

https://github.com/mne-tools/mne-bids-pipeline/actions/runs/6733693944/job/18302967345?pr=807

Could you help me out here, please?

File: mne_bids_pipeline/tests/configs/config_ERP_CORE.py [hunk 4]
```diff
@@ -71,13 +71,18 @@
 t_break_annot_start_after_previous_event = 3.0
 t_break_annot_stop_before_next_event = 1.5
 
-ica_reject = dict(eeg=350e-6, eog=500e-6)
-reject = "autoreject_global"
+if task == "N400":
```
Comment (hoechenberger) [inline mne_bids_pipeline/tests/configs/config_ERP_CORE.py:74]:
```suggestion
if task == "N400":  # test autoreject local without ICA
```

**Comment by hoechenberger** [comment]
Thanks @larsoner

File: mne_bids_pipeline/_config_import.py [hunk 5]
```diff
@@ -354,6 +354,7 @@ def _default_factory(key, val):
         {"custom": (8, 24.0, 40)},  # decoding_csp_freqs
         {"suffix": "ave"},  # source_info_path_update
         ["evoked"],  # inverse_targets
+        [1, 4, 8, 16],  # autoreject_n_interpolate
```
Comment (hoechenberger) [inline mne_bids_pipeline/_config_import.py:357]:
```suggestion
        [4, 8, 16],  # autoreject_n_interpolate
```

**Comment by hoechenberger** [comment]
Tests currently failing because of https://github.com/autoreject/autoreject/issues/320

**Comment by hoechenberger** [comment]
@larsoner I believe the CI error may have to do with https://github.com/mne-tools/mne-python/pull/12156
These errors did not occurs before that PR was merged :(

**Comment by hoechenberger** [comment]
@agramfort @larsoner 
Output looks like this:

https://output.circle-artifacts.com/output/job/26a8def0-8c16-44f2-917c-a197c07c406c/artifacts/0/reports/ERP_CORE/sub-015_ses-N400_task-N400_report.html#Epochs__Autoreject_cleaning

I think it's nice! Only quite a bit of whitespace around the figure. Is this something we should look to fix upstream in autoreject?

File: mne_bids_pipeline/steps/preprocessing/_08_ptp_reject.py [hunk 6]
```diff
@@ -79,46 +79,75 @@ def drop_ptp(
...
+            == reject_log.bad_epochs.sum()
+        )
+
+        msg = (
+            f"autoreject marked {reject_log.bad_epochs.sum()} epochs as bad "
+            f"(cross-validated n_interpolate limit: {ar.n_interpolate_})"
+        )
+        logger.info(**gen_log_kwargs(message=msg))
+    else:
+        reject = _get_reject(
```
Comment (hoechenberger) [inline mne_bids_pipeline/steps/preprocessing/_08_ptp_reject.py:110]: The diff doesn't look nice here, but I basically just added this new `if` block and moved the previous code into the `else` section

**Comment by hoechenberger** [comment]
Output looks like this now:
https://output.circle-artifacts.com/output/job/801b300e-8885-4e8c-a6e0-61648f3d69f0/artifacts/0/reports/ERP_CORE/sub-015_ses-N400_task-N400_report.html#Epochs__Autoreject_cleaning

I just restarted CI to take advantage of https://github.com/mne-tools/mne-python/pull/12167

**Comment by hoechenberger** [comment]
We're green!

File: docs/source/v1.5.md.inc [hunk 7]
```diff
@@ -15,6 +15,10 @@ All users are encouraged to update.
 - Input validation has been improved by leveraging [pydantic](https://docs.pydantic.dev) (#779 by @larsoner)
 - Reduced logging when reports are created and saved (#799 by @hoechenberger)
 - Added [`"picard-extended_infomax"`][mne_bids_pipeline._config.ica_algorithm] ICA algorithm to perform "extended Infomax"-like ICA decomposition using Picard (#801 by @hoechenberger)
+- Added support for using "local" [`autoreject`](https://autoreject.github.io) to find (and repoair) bad channels on a
```
Comment (hoechenberger) [inline docs/source/v1.5.md.inc:18]:
```suggestion
- Added support for using "local" [`autoreject`](https://autoreject.github.io) to find (and repair) bad channels on a
```

File: mne_bids_pipeline/steps/preprocessing/_08_ptp_reject.py [hunk 8]
```diff
@@ -79,46 +79,75 @@ def drop_ptp(
...
-        ch_types=cfg.ch_types,
-        param="reject",
-        epochs=epochs,
-    )
-    if cfg.spatial_filter == "ica":
-        ica_reject = _get_reject(
+
+    if cfg.reject == "autoreject_local":
+        import numpy as np
+        import autoreject
```
[the remark this answers is not in the document]
Comment (hoechenberger) [reply mne_bids_pipeline/steps/preprocessing/_08_ptp_reject.py:85]: We already depend upon autoreject because we've been using it for PTP rejection before ICA / SSP for a while :)

File: mne_bids_pipeline/steps/preprocessing/_08_ptp_reject.py [hunk 9]
```diff
@@ -79,46 +79,75 @@ def drop_ptp(
...
-        reject=cfg.reject,
-        ch_types=cfg.ch_types,
-        param="reject",
-        epochs=epochs,
-    )
-    if cfg.spatial_filter == "ica":
-        ica_reject = _get_reject(
+
+    if cfg.reject == "autoreject_local":
+        import numpy as np
```
[the remark this answers is not in the document]
Comment (hoechenberger) [reply mne_bids_pipeline/steps/preprocessing/_08_ptp_reject.py:84]: Good idea

File: mne_bids_pipeline/steps/preprocessing/_08_ptp_reject.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (larsoner) [reply mne_bids_pipeline/steps/preprocessing/_08_ptp_reject.py:85]: Okay then this can also be un-nested

[the remark this answers is not in the document]
Comment (hoechenberger) [reply mne_bids_pipeline/steps/preprocessing/_08_ptp_reject.py]: Done

**Comment by hoechenberger** [comment]
@larsoner I've addressed your comments

**Review by larsoner — APPROVED** [review]
